File: 算法/001-500/129.求根节点到叶节点数字之和/129.求根节点到叶节点数字之和.java
/*
 * @lc app=leetcode.cn id=129 lang=java
 *
 * [129] 求根节点到叶节点数字之和
 * 
 * 108/108 cases passed (7 ms)
 * Your runtime beats 5.43 % of java submissions
 * Your memory usage beats 5.05 % of java submissions (39.7 MB)
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int ans = 0;
    public int sumNumbers(TreeNode root) {
        travel(root, "");
        return ans;
    }

    private void travel(TreeNode root, String s) {
        if (root == null) return;
        s += root.val;
        if (root.left == null && root.right == null) {
            ans += Integer.parseInt(s);
            return;
        }
        travel(root.left, s);
        travel(root.right, s);
    }
}
// @lc code=end

File: 算法/001-500/129.求根节点到叶节点数字之和/129.求根节点到叶节点数字之和.ts
/*
 * @lc app=leetcode.cn id=129 lang=typescript
 *
 * [129] 求根节点到叶节点数字之和
 * 
 * 108/108 cases passed (60 ms)
 * Your runtime beats 89.22 % of typescript submissions
 * Your memory usage beats 16.17 % of typescript submissions (44 MB)
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */


function sumNumbers(root: TreeNode | null): number {
  let ans = 0
  function travel(root: TreeNode | null, prev: number) {
    if (!root) return
    let s = prev * 10 + root.val
    if (!root.left && !root.right) {
      ans += s
    } else {
      travel(root.left, s);
      travel(root.right, s);
    }
  }
  travel(root, 0)
  return ans
};
// @lc code=end

